File: python/103-Binary Tree Zigzag Level Order Traversal.py
from typing import List
from bintree_helper import *

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        "BFS with Reverse, Time: O(n), Space: O(n)"
        from collections import deque
        if root is None: return []
        q = deque([root])
        level = 0
        result = []
        while q:
            sz = len(q)
            result.append([])
            for _ in range(sz):
                node = q.popleft()
                result[-1].append(node.val)
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            if level%2 != 0:
                result[-1] = result[-1][::-1]
            level += 1
        return result
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        "BFS without Reverse, Time: O(n), Space: O(n)"
        from collections import deque
        if root is None: return []
        q = deque([root])
        level = 0
        result = []
        while q:
            sz = len(q)
            result.append([0]*sz)
            for i in range(sz):
                node = q.popleft()
                if level%2 == 0: idx = i
                else: idx = sz-i-1
                result[-1][idx] = node.val
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            level += 1
        return result

root = deserialize("[3,9,20,null,null,15,7]")
result = Solution().zigzagLevelOrder(root)
assert result == [[3], [20,9], [15,7]]

File: cpp/103. Binary Tree Zigzag Level Order Traversal.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return vvi();       
        queue<TreeNode*> q; q.push(root);

        int level = 1;
        vvi ans;
        while(q.size()) {
            bool fromLeft = (level++)&1;
            int qSize = q.size();
            ans.push_back(vi(qSize));
            vi &tmp = ans.back();
            for(int i = 0; i < qSize; i++) {
                TreeNode *node = q.front(); q.pop();
                if(fromLeft) tmp[i] = node->val;
                else tmp[qSize-1-i] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}